### ISSUE TITLE
Gradle: postprocess dependencies need a different model than bio-model

### ISSUE THREAD

**Issue by julie-sullivan** [intermine/intermine#1788]
@danielabutano 

`bio-postprocess` compiles and works correctly. However its tests actually store, retrieve and compare actual objects so requires the correct data model, e.g. `gene.setTranscript()`. `bio-postprocess` is testing FlyMine specific tasks so we need the FlyMine data model.

The correct solution to this is to have custom postprocessing for each mine. But maybe we can just get something working for now? Or should we just fix this properly?

Anyway in an attempt to fix, I updated `bio-postprocess` to depend on a FlyMine data source but that didn't work. I didn't get the compile errors anymore but had new problems:

1. Travis uses locally installed Gradle. Have to publish FlyMine's sources in JFrog too?
2. All the tests failed.

```
org.intermine.objectstore.ObjectStoreException: No ObjectStoreWriter properties were found for 'osw.bio-test'
	at org.intermine.objectstore.ObjectStoreWriterFactory.getObjectStoreWriter(ObjectStoreWriterFactory.java:52)
	at org.intermine.bio.postprocess.TransferSequencesTest.setUp(TransferSequencesTest.java:171)
	at junit.framework.TestCase.runBare(TestCase.java:132)
```
I had to remove the dependency on `bio-model` because `bio-model` has the wrong data model and I got conflicts.

Ideas on how I can go forward? This has caused Travis to fail.

**Comment by julie-sullivan** [comment]
Here's an idea:

* Change postprocessing to be like bio-sources, a multiproject
* Each postprocess has its own directory with its own data model
* Each postprocess will be its own project and JAR
Replace `PostProcessOperationsTask`? Or call it for every postprocess. 

Postprocesses sometimes call specific methods, e.g. `cl.setMissingChromosomeLocations()` for `set-missing-chromosome-locations`.

**Comment by julie-sullivan** [comment]
Problem we are trying to solve:

1. testCompile

The postprocesses themselves don't require a specific model to compile, only at runtime when it uses the model from the mine (I guess?). e.g. `PopulateChildFeature` operates on `InterMineObjects`.

The tests DO require operations on specific classes, e.g. `Gene.setExons()`. We used to use the biotestmine for this I think but we've moved that now.

2. @sammyjava has his own postprocesses. They are mine specific though. 

See #1351 and #1510

**Comment by danielabutano** [comment]
yes. I agree.
each postprocess can add its own dependencies (on bio-source which it depends on)

**Comment by julie-sullivan** [comment]
@danielabutano Here's what I did:

1. Moved common postprocess Util class to bio-core
2. changed postprocess to be a multiproject, removed `bio-postprocess` from `bio`
3. temporarily removed the rest of the postprocesses, just did `create-chromosome-lengths`
4. each postprocess is a project, has its own deps (solves the model issue with tests) and a property file containing the Java class to execute

What do you think? Should I go ahead with the rest?

* have to put these JARs on JFrog. I might need help with that.
* This allows Sam to define his own postprocesses.
* The postprocess runs but the test fails because it can't find `intermine.properties`. Do I need to make a specific `BioPostprocessDBModelPlugin`?

```
java.lang.RuntimeException: Could not load required global properties resource intermine.properties
	at org.intermine.util.PropertiesUtil.loadGlobalProperties(PropertiesUtil.java:149)
	at org.intermine.util.PropertiesUtil.initGlobalProperties(PropertiesUtil.java:54)
	at org.intermine.util.PropertiesUtil.getProperties(PropertiesUtil.java:41)
	at org.intermine.util.PropertiesUtil.getPropertiesStartingWith(PropertiesUtil.java:91)
	at org.intermine.objectstore.ObjectStoreWriterFactory.getObjectStoreWriter(ObjectStoreWriterFactory.java:50)
	at org.intermine.bio.postprocess.CreateChromosomeLocationsProcessTest.setUp(CreateChromosomeLocationsProcessTest.java:51)
```

**Comment by julie-sullivan** [comment]
Note that all postprocesses now have to extend `org.intermine.postprocess.PostProcessor`

**Comment by julie-sullivan** [comment]
@danielabutano The above solves one issue but still `bio-core`'s GFF3 test is creating `BindingSites`. Binding sites are not in the core model and I'm getting a travis error. There is no core data source that loads these data so I'll have to use a FlyMine one.

Should I push the FlyMine sources to JFrog / JCenter? Or can I just have Travis (and whoever wants to run tests) install FlyMine bio sources?

**Comment by sammyjava** [comment]
@julie-sullivan Actually a lot of my custom post-processors are generic, like one that fills in Publications from PubMed and CrossRef (so I don't have to build them in all my sources, just reference them with a PMID or DOI or title); etc. I just haven't bothered to push them on the IM community. :)

**Comment by julie-sullivan** [comment]
@sammyjava You know if you push your JARs to JCenter like we do, you can make them available to the IM community! :D 

Which makes me think we need some sort of registry so people can find others' sources ...

**Comment by sammyjava** [comment]
@julie-sullivan yeah, that's the registry of sources and displayers and widgets that we talked about a year ago....

**Comment by sammyjava** [comment]
And this is a most excellent change, which fixes the old issue of having to hack core post-processors because the layout wasn't built like sources. I've been using symlinks and stuff, very icky. Looking forward to this new layout!

**Comment by julie-sullivan** [comment]
@sammyjava I am glad! I've got it working for two postprocesses so far, just working through the rest.

I'm on a different branch here: https://github.com/julie-sullivan/intermine/tree/gradle-postprocess/bio/postprocess

1. Each postprocess has its own directory, like a source does
2. each postprocess now extends `PostProcessor`. The `postProcess()` method is called during the build.
3. each postprocess has its own dependencies. This will solve my problem with the `createUTR` tests not compiling.
4. You will be able to add your own postprocesses just like you add your own sources. 

Let me know what you think!

@danielabutano thought there might be a way to only have one `bio-postprocess.jar` instead of one JAR per postprocess. This would be preferable because we do have a LOT of JARs! However I couldn't figure out a way to have a single project AND still let you be able to add mine-specific processes AND solve my problem with each process needing different dependencies.

**Comment by julie-sullivan** [comment]
Fixed by #1794
